### ISSUE TITLE
[4.2 beta] - Error: Unknown dialect undefined on `strapi().load()`

### ISSUE THREAD

**Issue by boazpoolman** [strapi/strapi#12894]
### Describe the bug

When using the `4.2.0-beta.0` with typescript I get an error when I programatically load Strapi like this:
```
const strapi = require('@strapi/strapi');
// ...
const app = await strapi().load();
```

### Steps to reproduce the behavior

1. `npx create-strapi-app@beta test-app-beta --ts` use quickstart method
2. `cd test-app-beta`
3. `yarn add strapi-plugin-config-sync`
4. `yarn config-sync diff`
5. See error

### Expected behavior

Strapi to load correctly

### Code snippets (the error)

```
(node:96471) UnhandledPromiseRejectionWarning: Error: Unknown dialect undefined
    at getDialectClass (/Users/boazpoolman/Code/boazpoolman/strapi/test-app-beta/node_modules/@strapi/database/lib/dialects/index.js:12:13)
    at getDialect (/Users/boazpoolman/Code/boazpoolman/strapi/test-app-beta/node_modules/@strapi/database/lib/dialects/index.js:19:23)
    at new Database (/Users/boazpoolman/Code/boazpoolman/strapi/test-app-beta/node_modules/@strapi/database/lib/index.js:38:20)
    at Function.Database.init (/Users/boazpoolman/Code/boazpoolman/strapi/test-app-beta/node_modules/@strapi/database/lib/index.js:84:33)
    at Strapi.bootstrap (/Users/boazpoolman/Code/boazpoolman/strapi/test-app-beta/node_modules/@strapi/strapi/lib/Strapi.js:383:30)
    at Strapi.load (/Users/boazpoolman/Code/boazpoolman/strapi/test-app-beta/node_modules/@strapi/strapi/lib/Strapi.js:446:16)
    at async Command.<anonymous> (/Users/boazpoolman/Code/boazpoolman/strapi/test-app-beta/node_modules/strapi-plugin-config-sync/server/cli.js:220:17)
    at async Command.parseAsync (/Users/boazpoolman/Code/boazpoolman/strapi/test-app-beta/node_modules/strapi-plugin-config-sync/node_modules/commander/lib/command.js:923:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:96471) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:96471) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

### System

- Node.js version: v14.19.0
- Strapi version: 4.2.0-beta.0
- Database: sqlite
- Operating system: Mac OS (M1)

### Additional context

When I do an non-typescript install of the beta this issue does not happen.

**Comment by AntoineXevlabs** [comment]
Same thing encountered here. 

Seems like in `dialects.js`, `db.config.connection` is equal to `{}` (nothing is passed to the db config)

```
const getDialect = db => {
  const { client } = db.config.connection;
  console.log(db.config)
  const constructor = getDialectClass(client);
  const dialect = new constructor(db);
  dialect.client = client;

  return dialect;
};
```

**Comment by AntoineXevlabs** [comment]
@boazpoolman 
@soupette ⚠️ found the solution: you'll need to run `strapi start dist`
I think it's the typescript command start which is not well configured, a solution might be to change start command when generating with typescript flag `"start": "strapi start dist"`

**Comment by boazpoolman** [comment]
@AntoineXevlabs But that still doesn't fix the issue for me as I'm trying to start Strapi programatically, not via the CLI.

**Comment by AntoineXevlabs** [comment]
@boazpoolman change in package.json "start": "strapi start dist", then yarn / npm run start should work fine no ?

**Comment by boazpoolman** [comment]
@AntoineXevlabs Yeah, but what if you do this:

- Create file `test.js`
- Add to package.json: `"main": "./test.js"`
- Inside `test.js` add:
```
const strapi = require('@strapi/strapi');
// ...
const app = await strapi().load();
``` 
- Run `node .`
- See the error

**Comment by AntoineXevlabs** [comment]
@boazpoolman `const app = strapi({appDir: './dist'}).load();` should make the trick

**Comment by boazpoolman** [comment]
Awesome! That seems legit.

Question for Strapi;
Is this backwards compatible?
Should I update my plugins to this format?

**Comment by Convly** [comment]
Hey, thanks for opening this issue!

Could you try using `const app = strapi({ distDir: './dist' }).load();` instead and tell me it works?

Basically, for the moment the dist directory represents the compiled project (with the same folder structure as the app directory) and the app directory represents your codebase (TS or JS).
The main differences are:
- When using the CTB to create/update/delete content types, we're using the app folder to write the files. When reading what content types exist on your system, we're reading the dist folder.
- Only the app folder is watched when using the develop command. In TS a compilation will be triggered when a change is detected in the app folder & the output will be written to the dist folder

Note: We're having some issues with the public folder as it should be static and shouldn't be compiled/moved in any way, which isn't the case for the moment.

Note 2: When developing in vanilla JS, the dist folder will be the same as the app folder since we writing & reading in the same place. If we were to use some kind of compilation tool for JS (babel, etc...) we would set the dist directory to wherever babel output the code.

Note 3: Currently, the default values for the app & dist directories are set using this small util.
```javascript
/**
 * Resolve the working directories based on the instance options.
 *
 * Behavior:
 * - `appDir` is the directory where Strapi will write every file (schemas, generated APIs, controllers or services)
 * - `distDir` is the directory where Strapi will read configurations, schemas and any compiled code
 *
 * Default values:
 * - If `appDir` is `undefined`, it'll be set to `process.cwd()`
 * - If `distDir` is `undefined`, it'll be set to `appDir`
 */
const resolveWorkingDirectories = opts => {
  const cwd = process.cwd();

  const appDir = opts.appDir ? path.resolve(cwd, opts.appDir) : cwd;
  const distDir = opts.distDir ? path.resolve(cwd, opts.distDir) : appDir;

  return { appDir, distDir };
};

```

**Comment by boazpoolman** [comment]
Hey @Convly,

Thanks for following up on this issue.

I find that using either of these two give me the expected result:

```
const app = strapi({ distDir: './dist' }).load();
const app = strapi({ appDir: './dist' }).load();
```

Though if I leave the `strapi()` parameters empty I will get the error as initially described.

**Comment by jobehi** [comment]
> @boazpoolman
> @soupette ⚠️ found the solution: you'll need to run `strapi start dist`
> I think it's the typescript command start which is not well configured, a solution might be to change start command when generating with typescript flag `"start": "strapi start dist"`

Thank you that worked for me.

**Comment by Convly** [comment]
> @boazpoolman @soupette warning found the solution: you'll need to run `strapi start dist` I think it's the typescript command start which is not well configured, a solution might be to change start command when generating with typescript flag `"start": "strapi start dist"`

Starting from v4.2.0-beta.2, you should be able to only use `strapi start` instead of `strapi start dist`. I'll close this issue as it has been answered.

Please note that we're also working on providing better errors when Strapi has been started without the correct directories.

**Comment by boazpoolman** [comment]
Hey @Convly

I think the issue is misunderstood here.
I don't have issues when running strapi from the cli with `strapi start`.

The issue I'm having is when I try to start Strapi programatically from a js file like this:

```
const strapi = require('@strapi/strapi');
// ...
const app = strapi().load();
```

Which still gives me the same error on the latest `v4.2.0-beta.3` release.

**Comment by Convly** [comment]
Hey @boazpoolman,

When instantiating Strapi programmatically:

### Neither `appDir` or `distDir` are passed

Both `distDir` and `appDir` will be set to `process.cwd()`

### Only `appDir` is defined

`distDir` value will match the one for `appDir`.

### Only `distDir` is defined

`appDir` will be set to `process.cwd()` and `distDir` will keep its given value

So basically, if you've compiled code stored in a specific directory (eg: when using TypeScript) you should instantiate Strapi with a specific `distDir` value which match the path of your build directory.

However you shouldn't set the `appDir` to your build/dist folder as it could cause issues when the app will try to write some files.

**Comment by Convly** [comment]
Do you think it would be worth making dedicated documentation for how to instantiate Strapi programmatically, especially with TypeScript? I can see how it can cause some confusion.

**Comment by boazpoolman** [comment]
Yes. That makes sense to me, as I'm still confused 🤦

I have[ a plugin ](https://github.com/boazpoolman/strapi-plugin-config-sync) which could be used by someone using Strapi with Typescript, or without.

In the plugin I have [a CLI that starts Strapi programatically](https://github.com/boazpoolman/strapi-plugin-config-sync/blob/master/server/cli.js#L72).

Should I pass the distDir there? And will that then still work for people using Strapi without Typescript?

**Comment by Convly** [comment]
Nice!

Also, please note that the incremental option might be faulty at the moment (it basically ignore the option), I'll work on that later.

**Comment by boazpoolman** [comment]
Just wanted to let you know that this worked! :)

**Comment by Convly** [comment]
Hey @boazpoolman :wave: 

Glad to hear it worked!

On my side, I just wanted to share that we've released the `4.3.0-beta.1` (`@beta` tag) including an easier way to start strapi programmatically while handling potential JS or TS applications.

You can now use the following snippet to create an instance of Strapi + compile the code if needed:
```js
const strapi = require('@strapi/strapi');

const appContext = await strapi.compile(); // { appDir: string; distDir: string; }
const app = await strapi(appContext).load();
```
